core/src/test/java/com/linecorp/armeria/server/HttpServerAbortingInfiniteStreamTest.java

```diff
@@ -0,0 +1,135 @@
+ */
+package com.linecorp.armeria.server;
+
+import static com.linecorp.armeria.common.SessionProtocol.H1C;
+import static com.linecorp.armeria.common.SessionProtocol.H2C;
+import static org.assertj.core.api.Assertions.assertThat;
+import static org.awaitility.Awaitility.await;
+
+import java.util.Collection;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicBoolean;
+
+import javax.annotation.Nullable;
+
+import org.junit.Rule;
+import org.junit.Test;
+import org.junit.runner.RunWith;
+import org.junit.runners.Parameterized;
+import org.junit.runners.Parameterized.Parameters;
+import org.reactivestreams.Subscriber;
+import org.reactivestreams.Subscription;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import com.google.common.collect.ImmutableList;
+
+import com.linecorp.armeria.client.HttpClient;
+import com.linecorp.armeria.common.HttpData;
+import com.linecorp.armeria.common.HttpHeaders;
+import com.linecorp.armeria.common.HttpMethod;
+import com.linecorp.armeria.common.HttpObject;
+import com.linecorp.armeria.common.HttpResponse;
+import com.linecorp.armeria.common.HttpResponseWriter;
+import com.linecorp.armeria.common.HttpStatus;
+import com.linecorp.armeria.common.SessionProtocol;
+import com.linecorp.armeria.testing.server.ServerRule;
+
+@RunWith(Parameterized.class)
+public class HttpServerAbortingInfiniteStreamTest {
+    private static final Logger logger = LoggerFactory.getLogger(HttpServerAbortingInfiniteStreamTest.class);
+
+    @Parameters(name = "{index}: protocol={0}")
+    public static Collection<SessionProtocol> protocols() {
+        return ImmutableList.of(H1C, H2C);
+    }
+
+    private final SessionProtocol protocol;
+    private final AtomicBoolean isCompleted = new AtomicBoolean();
+
+    public HttpServerAbortingInfiniteStreamTest(SessionProtocol protocol) {
+        this.protocol = protocol;
+    }
+
+    @Rule
+    public final ServerRule server = new ServerRule() {
+        @Override
+        protected void configure(ServerBuilder sb) throws Exception {
+            sb.service("/infinity", (ctx, req) -> {
+                // Ensure that the protocol is expected one.
+                assertThat(ctx.sessionProtocol()).isEqualTo(protocol);
+
+                final HttpResponseWriter writer = HttpResponse.streaming();
+                writer.write(HttpHeaders.of(HttpStatus.OK));
+
+                // Do not close the response writer because it returns data infinitely.
+                writer.onDemand(new Runnable() {
+                    @Override
+                    public void run() {
+                        writer.write(HttpData.ofUtf8("infinite stream"));
+                        writer.onDemand(this);
+                    }
+                });
+                writer.completionFuture().whenComplete((unused, cause) -> {
+                    // We are not expecting that this stream is successfully finished.
+                    if (cause != null) {
```

Comment:
I am curious if we need to assert the cause is what we expected.

Reply:
Added the following code:
```
if (protocol == H1C) {
    assertThat(cause).isInstanceOf(CancelledSubscriptionException.class);
} else {
    assertThat(cause).isInstanceOf(AbortedStreamException.class);
}
```